src/base/Gas.ml

```diff
@@ -104,70 +104,157 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
+      match ee with
+      | Literal _ | Var _ | Message _ | App _ | Constr _ | TApp _ ->
+          GasExpr (GasCharge.StaticCost 1, e)
+      | Fixpoint (f, t, e') ->
+          GasExpr
+            ( GasCharge.StaticCost 1,
+              (Fixpoint (f, t, expr_static_cost e'), erep) )
+      | Fun (f, t, e') ->
+          GasExpr
+            (GasCharge.StaticCost 1, (Fun (f, t, expr_static_cost e'), erep))
+      | TFun (f, e') ->
+          GasExpr (GasCharge.StaticCost 1, (TFun (f, expr_static_cost e'), erep))
+      | Let (i, t, lhs, rhs) ->
+          GasExpr
+            ( GasCharge.StaticCost 1,
+              (Let (i, t, expr_static_cost lhs, expr_static_cost rhs), erep) )
+      | MatchExpr (o, clauses) ->
+          GasExpr
+            ( GasCharge.StaticCost (List.length clauses),
+              ( MatchExpr
+                  ( o,
+                    List.map clauses ~f:(fun (p, e') ->
+                        (p, expr_static_cost e')) ),
+                erep ) )
+      | Builtin _
+      (* We don't add costs for Builtin because we can't know it statically
+       * without type info (Eval doesn't run the type checker). To know this
+       * statically, call builtin_cost below, providing argument types. *)
+      | GasExpr _ ->
+          ee
+    in
+    (e', erep)
 
   (* this is a dynamic cost. *)
+  let rec stmts_cost = function
+    | [] -> []
+    | (s, srep) :: rem_stmts ->
+        let s' =
+          match s with
+          | Load (x, _) ->
+              let g =
+                GasStmt
+                  (GasCharge.SumOf
+                     ( GasCharge.SizeOf (GI.get_id x),
+                       GasCharge.MapSortCost (GI.get_id x) ))
+              in
+              (* We charge *after* the load because we can't know the size before. *)
+              [ (s, srep); (g, srep) ]
+          | Store (_, v) | SendMsgs v | CreateEvnt v ->
+              let g = GasStmt (GasCharge.SizeOf (GI.get_id v)) in
+              [ (g, srep); (s, srep) ]
+          | Bind (x, e) ->
+            let g = GasStmt (GasCharge.StaticCost 1) in
+            let s' = Bind (x, expr_static_cost e) in
+            [ (g, srep); (s', srep) ]
+          | ReadFromBC _ | CallProc _ ->
+              let g = GasStmt (GasCharge.StaticCost 1) in
+              [ (g, srep); (s, srep) ]
+          | MapUpdate (_, klist, ropt) ->
+              let n = GasCharge.StaticCost (List.length klist) in
+              let g =
+                match ropt with
+                | Some r -> GasCharge.SumOf (GasCharge.SizeOf (GI.get_id r), n)
+                | None -> n
+              in
+              [ (GasStmt g, srep); (s, srep) ]
+          | MapGet (x, _, klist, _) ->
+              let n = GasCharge.StaticCost (List.length klist) in
+              let g =
+                  GasCharge.SumOf
+                    ( GasCharge.SumOf
+                        ( GasCharge.SizeOf (GI.get_id x),
+                          GasCharge.MapSortCost (GI.get_id x) ),
```

Comment:
Because when we fetch a Map value from the blockchain, the protobuf transmission doesn't guarantee any order on key-val pairs (i.e., maps). (Note that leveldb storage is ordered, so direct key accesses are still efficient). So to guarantee an order for Scilla, we sort map values when bringing them in. That's the cost.

Reply:
But why do we need to order the map on the Scilla side? Don't we just drop it all down a hashmap?

Reply:
After referring back to https://github.com/Zilliqa/scilla/pull/764, I recall that we had an issue with reaching consensus because of this. With a non-deterministic order, two different Scilla executions, on different nodes, can lead to different outputs.

Reply:
Oh, right - there was something like that.

It looks like Jacob in January knew what he was talking about, so who is Jacob in October to question that...